source/extensions/filters/http/json_to_metadata/filter.h

```diff
@@ -0,0 +1,136 @@
+using KeyValuePair =
+    envoy::extensions::filters::http::json_to_metadata::v3::JsonToMetadata::KeyValuePair;
+using ValueType = envoy::extensions::filters::http::json_to_metadata::v3::JsonToMetadata::ValueType;
+
+/**
+ * Data structure to store one rule.
+ */
+struct Rule {
+  Rule(const ProtoRule& rule);
+  const ProtoRule rule_;
+  std::vector<std::string> keys_;
+};
+
+using Rules = std::vector<Rule>;
+
+/**
+ * Configuration for the Json to Metadata filter.
+ */
+class FilterConfig {
+public:
+  FilterConfig(
+      const envoy::extensions::filters::http::json_to_metadata::v3::JsonToMetadata& proto_config,
+      Stats::Scope& scope);
+  ~FilterConfig() = default;
+
+  JsonToMetadataStats& stats() { return stats_; }
+  // True if we have rules for requests
+  bool doRequest() const { return !request_rules_.empty(); }
+  const Rules& requestRules() const { return request_rules_; }
+  uint32_t requestBufferLimitBytes() const { return request_buffer_limit_bytes_; }
+  bool requestContentTypeAllowed(absl::string_view) const;
+
+private:
+  using ProtobufRepeatedRule = Protobuf::RepeatedPtrField<ProtoRule>;
+  Rules generateRequestRules(
+      const envoy::extensions::filters::http::json_to_metadata::v3::JsonToMetadata& proto_config);
+  absl::flat_hash_set<std::string> generateRequestAllowContentTypes(
+      const envoy::extensions::filters::http::json_to_metadata::v3::JsonToMetadata& proto_config);
+  JsonToMetadataStats stats_;
+  const uint32_t request_buffer_limit_bytes_;
+  const Rules request_rules_;
+  const absl::flat_hash_set<std::string> request_allow_content_types_;
+  const bool request_allow_empty_content_type_;
+};
+
+const uint32_t MAX_PAYLOAD_VALUE_LEN = 8 * 1024;
+
+/**
+ * HTTP Json to Metadata Filter.
+ */
+class Filter : public Http::PassThroughFilter, Logger::Loggable<Logger::Id::filter> {
+public:
+  Filter(std::shared_ptr<FilterConfig> config) : config_(config){};
+  ~Filter() override;
+
+  Http::FilterHeadersStatus decodeHeaders(Http::RequestHeaderMap& headers,
+                                          bool end_stream) override;
+  Http::FilterDataStatus decodeData(Buffer::Instance& data, bool end_stream) override;
+  Http::FilterTrailersStatus decodeTrailers(Http::RequestTrailerMap&) override;
+
+private:
+  using StructMap = absl::flat_hash_map<std::string, ProtobufWkt::Struct>;
+  // Handle on_missing case of the `rule` and store in `struct_map`.
+  void handleOnMissing(const Rule& rule, StructMap& struct_map);
+  // Handle on_present case of the `rule` and store in `struct_map`, which depends on
+  // the value of `parent_node->key`.
+  absl::Status handleOnPresent(Json::ObjectSharedPtr parent_node, const std::string& key,
+                               const Rule& rule, StructMap& struct_map);
+
+  // Process the case without body, i.e., on_missing is applied for all rules.
+  void handleAllOnMissing(const Rules& rules, bool& reported_flag);
```

Comment:
can you provide insight on why we pass around the reported_flag which is just the `request_processing_finished_` flag?

Reply:
in the future we might handle response as well, so have a `out` variable. I'll rename it to `processing_finished_flag`